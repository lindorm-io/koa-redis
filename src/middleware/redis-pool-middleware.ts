import { Middleware } from "@lindorm-io/koa";
import { RedisConnection } from "@lindorm-io/redis";
import { RedisContext } from "../types";

export const redisPoolMiddleware =
  (redis: RedisConnection): Middleware<RedisContext> =>
  async (ctx, next): Promise<void> => {
    const metric = ctx.getMetric("redis");

    if (!redis.isConnected()) {
      await redis.connect();
    }

    ctx.client.redis = redis;

    ctx.logger.debug("redis connection added to context");

    metric.end();

    await next();
  };
