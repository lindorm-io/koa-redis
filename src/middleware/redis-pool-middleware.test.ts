import MockDate from "mockdate";
import { RedisConnection, RedisConnectionType } from "@lindorm-io/redis";
import { redisPoolMiddleware } from "./redis-pool-middleware";
import { logger } from "../test";
import { Metric } from "@lindorm-io/koa";

jest.mock("uuid", () => ({
  v4: () => "e397bc49-849e-4df6-a536-7b9fa3574ace",
}));

MockDate.set("2020-01-01T08:00:00.000Z");

const next = jest.fn();

describe("redisPoolMiddleware", () => {
  let ctx: any;
  let inMemoryCache: Record<string, any>;
  let redis: RedisConnection;

  beforeEach(() => {
    inMemoryCache = { initialized: true };

    redis = new RedisConnection({
      type: RedisConnectionType.MEMORY,
      port: 100,
      inMemoryCache,
    });

    ctx = {
      client: {},
      logger,
      metrics: {},
    };
    ctx.getMetric = (key: string) => new Metric(ctx, key);
  });

  test("should set a functional redis on context", async () => {
    await redis.connect();

    await expect(redisPoolMiddleware(redis)(ctx, next)).resolves.toBeUndefined();

    expect(ctx.client.redis).toStrictEqual(expect.any(RedisConnection));
    expect(ctx.metrics.redis).toStrictEqual(expect.any(Number));

    const client = ctx.client.redis.client();
    await expect(client.set("key", { blob: "yes" })).resolves.toBe("OK");
    await expect(client.get("key")).resolves.toMatchSnapshot();

    expect(inMemoryCache).toMatchSnapshot();
  });

  test("should automatically connect", async () => {
    await expect(redisPoolMiddleware(redis)(ctx, next)).resolves.toBeUndefined();

    expect(ctx.client.redis).toStrictEqual(expect.any(RedisConnection));
    expect(ctx.metrics.redis).toStrictEqual(expect.any(Number));

    const client = ctx.client.redis.client();
    await expect(client.set("key", { blob: "yes" })).resolves.toBe("OK");
    await expect(client.get("key")).resolves.toMatchSnapshot();

    expect(inMemoryCache).toMatchSnapshot();
  });
});
