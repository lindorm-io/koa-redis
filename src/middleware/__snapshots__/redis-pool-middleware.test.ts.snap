// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`redisPoolMiddleware should automatically connect 1`] = `
Object {
  "blob": "yes",
}
`;

exports[`redisPoolMiddleware should automatically connect 2`] = `
Object {
  "initialized": true,
  "key": Object {
    "blob": "{\\"json\\":{\\"blob\\":\\"yes\\"},\\"meta\\":{\\"blob\\":\\"string\\"}}",
    "expiresInSeconds": undefined,
  },
}
`;

exports[`redisPoolMiddleware should set a functional redis on context 1`] = `
Object {
  "blob": "yes",
}
`;

exports[`redisPoolMiddleware should set a functional redis on context 2`] = `
Object {
  "initialized": true,
  "key": Object {
    "blob": "{\\"json\\":{\\"blob\\":\\"yes\\"},\\"meta\\":{\\"blob\\":\\"string\\"}}",
    "expiresInSeconds": undefined,
  },
}
`;
